cfgmgr/vrfmgr.cpp

```diff
@@ -0,0 +1,146 @@
+#include <string.h>
+#include "logger.h"
+#include "dbconnector.h"
+#include "producerstatetable.h"
+#include "tokenize.h"
+#include "ipprefix.h"
+#include "vrfmgr.h"
+#include "exec.h"
+#include "shellcmd.h"
+
+#define VRF_TABLE_START 100
+#define VRF_TABLE_END 164
+
+using namespace swss;
+
+VrfMgr::VrfMgr(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, const vector<string> &tableNames) :
+        Orch(cfgDb, tableNames)
+{
+    for (uint32_t i = VRF_TABLE_START; i < VRF_TABLE_END; i++)
+    {
+        m_freeTables.emplace(i);
+    }
+}
+
+uint32_t VrfMgr::getFreeTable(void)
+{
+    SWSS_LOG_ENTER();
+
+    if (m_freeTables.empty())
+    {
+        return 0;
+    }
+
+    uint32_t table = *m_freeTables.begin();
+    m_freeTables.erase(table);
+
+    return table;
+}
+
+void VrfMgr::returnTable(uint32_t table)
+{
+    SWSS_LOG_ENTER();
+
+    m_freeTables.emplace(table);
+}
+
+bool VrfMgr::delLink(const string& vrfName)
+{
+    SWSS_LOG_ENTER();
+
+    stringstream cmd;
+    string res;
+
+    if (m_vrfTableMap.find(vrfName) == m_vrfTableMap.end())
+    {
+        return false;
+    }
+
+    returnTable(m_vrfTableMap[vrfName]);
+
+    cmd << IP_CMD << " link del " << vrfName;
+    int ret = swss::exec(cmd.str(), res);
+    if (ret)
+    {
+        SWSS_LOG_ERROR("Failed command %s, rc:%d", cmd.str().c_str(), ret);
+        return false;
+    }
+
+    m_vrfTableMap.erase(vrfName);
+
+    return true;
+}
+
+bool VrfMgr::setLink(const string& vrfName)
+{
+    SWSS_LOG_ENTER();
+
+    stringstream cmd;
+    string res;
+
+    uint32_t table = getFreeTable();
+    if (table == 0)
+    {
+        return false;
+    }
+
```

Comment:
can we check the vrfTableMap, if it is there means the vrf has been created, then we can just return true. Instead of creating a new one.

Reply:
Done